server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -55,6 +96,73 @@ public TransportUnpromotableShardRefreshAction(
             transportService.getThreadPool().executor(ThreadPool.Names.REFRESH)
         );
         this.indicesService = indicesService;
+        this.threadPool = threadPool;
+        this.maxWaitTimeForRefreshUnblock = maxWaitTimeForRefreshUnblock;
+        this.useRefreshBlock = useRefreshBlock;
+    }
+
+    @Override
+    protected void beforeDispatchingRequestToUnpromotableShards(UnpromotableShardRefreshRequest request, ActionListener<Void> listener) {
+        if (useRefreshBlock == false) {
+            ActionListener.completeWith(listener, () -> null);
+            return;
+        }
+
+        new IndexRefreshUnBlockObserver(request.shardId().getIndexName(), listener).run();
+    }
+
+    class IndexRefreshUnBlockObserver extends AbstractRunnable {
+        private final String indexName;
+        private final ActionListener<Void> listener;
+        private final ClusterStateObserver clusterStateObserver;
+
+        IndexRefreshUnBlockObserver(String indexName, ActionListener<Void> listener) {
+            this.indexName = indexName;
+            this.listener = listener;
+            this.clusterStateObserver = new ClusterStateObserver(
+                clusterService,
+                maxWaitTimeForRefreshUnblock,
+                logger,
+                threadPool.getThreadContext()
+            );
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            var state = clusterStateObserver.setAndGetObservedState();
+            var indexLevelBlockException = isIndexBlockedOnRefresh(state);
+            if (indexLevelBlockException == null) {
+                ActionListener.completeWith(listener, () -> null);
+                return;
+            }
+
+            clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    run();
+                }
+
+                @Override
+                public void onClusterServiceClose() {
+                    listener.onFailure(new NodeClosedException(clusterService.localNode()));
+                }
+
+                @Override
+                public void onTimeout(TimeValue timeout) {
+                    listener.onFailure(indexLevelBlockException);
```

Comment:
I was wondering if we should just rely on the fact that the blocks would be cleared eventually instead of failing the refresh request, wdyt?

Reply:
I could see us do both. But in the case of bulks it would be nice to use the timeout from the bulk instead.

Also, I think throwing the refresh block exception is confusing, I'd prefer to throw a timeout exception instead.

But we could also forego it, expecting the refresh unblock to happen automatically after a separate timeout.

Reply:
For refresh triggered by writes we can maybe reuse the `postWriteRefreshTimeout` and for explicit refreshes reuse the timeout from the original request, and then rely on the block to be cleared automatically in case we wait indefinitely? And get rid of the new setting?

Reply:
That makes more sense, I used the `postWriteRefreshTimeout` instead of the setting in https://github.com/elastic/elasticsearch/pull/120642/commits/9a22ad6ae9825196fe8a452435a2f02dd90278d8